lms/djangoapps/courseware/module_render.py

```diff
@@ -846,8 +882,13 @@ def get_module_for_descriptor_internal(user, descriptor, student_data, course_id
     # that affects xblock visibility.
     user_needs_access_check = getattr(user, 'known', True) and not isinstance(user, SystemUser)
     if user_needs_access_check:
-        if not has_access(user, 'load', descriptor, course_id):
-            return None
+        access = has_access(user, 'load', descriptor, course_id)
+        if not access:
+            if access.user_message or access.user_fragment:
+                # This content will be access restricted by modifying the outgoing html
+                return descriptor
```

Comment:
We need to double-check the other possible code-paths that call this function and make sure they aren't assuming that the code returns a descriptor iff the user has access. For instance: https://github.com/edx/edx-platform/blob/master/lms/djangoapps/courseware/module_render.py#L869 is one such place that is using `None` to mean no-access.

One possible safety net is to add to the already gigantic raft of parameters to this function, but I'm not sure that's the best course of action.

(anything starting with `get_module` is suspect, btw).

Reply:
Fixed with flag.